tensorly/decomposition/tests/test_candecomp_parafac.py

```diff
@@ -1,10 +1,127 @@
 import numpy as np
-from ..candecomp_parafac import parafac
-from ..candecomp_parafac import non_negative_parafac
+import pytest
+
+from ..candecomp_parafac import (
+    parafac, non_negative_parafac, normalize_factors, initialize_factors)
 from ...kruskal_tensor import kruskal_to_tensor
 from ...random import check_random_state
 from ... import backend as T
 
+
+randomized_test_cases = {
+    'order3_small': {
+        'dimensions': (8, 8, 8), 'rank': 4, 'seed': 0},
+    'order3_large': {
+        'dimensions': (16, 16, 16), 'rank': 12, 'seed': 0},
+    'order3_mixed': {
+        'dimensions': (16, 12, 8), 'rank': 8, 'seed': 0},
+    'order3_small_orthogonal': {
+        'dimensions': (8, 8, 8), 'rank': 4, 'seed': 0, 'orthogonal': True},
+    'order3_large_orthogonal': {
+        'dimensions': (16, 16, 16), 'rank': 12, 'seed': 0, 'orthogonal': True},
+    'order3_mixed_orthogonal': {
+        'dimensions': (16, 12, 8), 'rank': 8, 'seed': 0, 'orthogonal': True},
+    'order4': {
+        'dimensions': (8, 8, 8, 8), 'rank': 4, 'seed': 0},
+    'order4_orthogonal': {
+        'dimensions': (8, 8, 8, 8), 'rank': 4, 'seed': 0, 'orthogonal': True},
+}
+
+
+def create_random_decomposition(dimensions=(10,10,10), rank=10, seed=None,
+                                orthogonal=False, noisy=False):
+    r"""Random tensor generation used in testing.
+
+    Returns a random tensor of specified order along with a list of its
+    constituent normalized factor matrices and corresponding weight vector.
+
+    Parameters
+    ----------
+    dimensions : tuple
+    rank : int
+        The desired rank of the tensor. Note that if `orthogonal == True` then
+        the rank must be less than or equal to the smallest dimension of the
+        tensor.
+    seed : int
+        Random number generation seed.
+    orthogonal : bool
+        If `True`, creates a tensor with orthogonal rank-1 components.
+    noisy : bool
+        If `True`, adds Gaussian noise to the tensor. (Useful when constructing
+        near-orthogonal tensors for experimentation.)
+
+    Returns
+    -------
+    tensor : ndarray
+    factors : list of 2D ndarrays
+        A list of random ALS factors.
+    weights : 1D ndarray
+        A list of random weights.
+
+    """
+    order = len(dimensions)
+    dim = min(dimensions)
+    if (rank > dim) and orthogonal:
+        raise ValueError('Can only construct orthogonal tensors when '
+                         'rank <= min(dimensions)')
+
+    np.random.seed(seed)
+    weights = 4*T.arange(1, rank+1)
+    factors = [T.tensor(np.random.randn(dim,rank)) for dim in dimensions]
+
+    if orthogonal:
+        factors = [T.qr(factor)[0] for factor in factors]
+    tensor = kruskal_to_tensor(factors, weights)
+
+    if noisy:
+        scale = rank/(rank+10.**(order-1))
+        tensor += T.tensor(scale*np.random.randn(*dimensions))
+    return tensor, factors, weights
+
+
+def test_parafac_errors():
```

Comment:
I would remove the parameterized test and have just one function.
Also i'd use numpy.testing.assert_raises rather than pytest (this way it still runs with nose).

Reply:
1. (See comments below regarding parametrized testing.)
2. Sure.